DataStructures/MinMaxHeap.cs

```diff
@@ -0,0 +1,399 @@
+        public bool RemoveMin()
+        {
+            if (Count == 0)
+            {
+                return false;
+            }
+
+            RemoveNode(0);
+            return true;
+        }
+
+        /// <summary>
+        /// Removes the maximum node from the heap and returns its value.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">Thrown if heap is empty.</exception>
+        /// <returns>Value of the removed maximum node.</returns>
+        public T ExtractMax()
+        {
+            if (Count == 0)
+            {
+                throw new InvalidOperationException("Heap is empty");
+            }
+
+            T max = GetMax();
+            RemoveNode(GetMaxNodeIndex());
+            return max;
+        }
+
+        /// <summary>
+        /// Removes the minimum node from the heap and returns its value.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">Thrown if heap is empty.</exception>
+        /// <returns>Value of the removed minimum node.</returns>
+        public T ExtractMin()
+        {
+            if (Count == 0)
+            {
+                throw new InvalidOperationException("Heap is empty");
+            }
+
+            T min = GetMin();
+            RemoveNode(0);
+            return min;
+        }
+
+        /// <summary>
+        /// Gets the maximum value in the heap, as defined by the comparer.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">Thrown if heap is empty.</exception>
+        /// <returns>The maximum value in the heap.</returns>
+        public T GetMax()
+        {
+            if (Count == 0)
+            {
+                throw new InvalidOperationException("Heap is empty");
+            }
+
+            return Heap[GetMaxNodeIndex()];
+        }
+
+        /// <summary>
+        /// Gets the minimum value in the heap, as defined by the comparer.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">Thrown if heap is empty.</exception>
+        /// <returns>The minimum value in the heap.</returns>
+        public T GetMin()
+        {
+            if (Count == 0)
+            {
+                throw new InvalidOperationException("Heap is empty");
+            }
+
+            return Heap[0];
+        }
+
+        /// <summary>
+        /// Finds maximum or minumum value among children and grandchildren of the specified node.
+        /// </summary>
+        /// <param name="index">Index of the node in the Heap array.</param>
+        /// <param name="minimum">If true find minimum value, otherwise - maximum.</param>
+        /// <returns>Index of the maximum or minimum descendant.</returns>
+        private int ExtremumChildOrGrandchild(int index, bool minimum)
+        {
+            var descendants = new int[] { 2 * index + 1, 2 * index + 2, 4 * index + 3, 4 * index + 4, 4 * index + 5, 4 * index + 6 };
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove the array type; it is redundant.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5421646)